File: RadarTools/src/shared/Shared/IO.hs
{-

Copyright 2017 Robert Christian Taylor

Licensed under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License.
You may obtain a copy of the License at

    http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software
distributed under the License is distributed on an "AS IS" BASIS,
WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
See the License for the specific language governing permissions and
limitations under the License.

-}

{- |
Module      :  Shared.IO
Description :  Serialization and deserialization code
Copyright   :  (c) Robert C. Taylor
License     :  Apache 2.0

Maintainer  :  r0wbrt@gmail.com
Stability   :  unstable
Portability :  portable

-}

module Shared.IO
    (
          serializeOutput
        , deserializeInput
        , complexFloatSerializer
        , complexSigned16SerializerOne
        , complexDoubleSerializer
        ) where




import qualified Data.Binary.Put        as BP
import qualified Data.ByteString        as B
import           Data.ByteString.Unsafe (unsafePackCStringFinalizer,
                                         unsafeUseAsCStringLen)
import           Data.Complex
import qualified Data.Vector.Storable   as VST
import           Data.Word
import           Foreign.C.String       (CString)
import           Foreign.ForeignPtr     (newForeignPtr, withForeignPtr)
import           Foreign.Marshal.Alloc  (finalizerFree, free, mallocBytes)
import           Foreign.Ptr            (Ptr, castPtr)
import           Foreign.Storable       (sizeOf)
import           GHC.Float
import qualified Shared.CommandLine     as CL
import           System.IO.Unsafe       (unsafeDupablePerformIO)



{- ### C Functions for serialization ### -}

-- | Converts a complex float array to an array of magnitudes.
foreign import ccall unsafe "convertComplexFloatArrayToDoubleMag" c_convertComplexFloatArrayToDoubleMag ::
        Ptr (Complex Float)
    ->  Ptr Double
    ->  Int
    ->  IO ()


-- | Converts a complex float array to an array of magnitudes stored as floats.
foreign import ccall unsafe "convertComplexFloatArrayToFloatMag" c_convertComplexFloatArrayToFloatMag ::
        Ptr (Complex Float)
    ->  Ptr Float
    ->  Int
    ->  IO ()


-- | Converts an array of complex float to an array to an array of their
--   magnitudes and stores them as a signed 16 values.
foreign import ccall unsafe "convertComplexFloatArrayToSigned16Mag" c_convertComplexFloatArrayToSigned16Mag ::
        Ptr (Complex Float)
    ->  Ptr Word16
    ->  Int
    ->  IO ()


-- | Makes a copy of the complex float array
foreign import ccall unsafe "convertFloatArrayToFloatArray" c_convertFloatArrayToFloatArray ::
        Ptr (Complex Float)
    ->  Ptr (Complex Float)
    ->  Int
    ->  IO ()


-- | Converts a complex float array to a complex signed 16 array using saturation
--  arithmetic.
foreign import ccall unsafe "convertComplexFloatArrayToComplexSigned16" c_convertComplexFloatArrayToComplexSigned16 ::
        Ptr (Complex Float)
    ->  Ptr Word16
    ->  Int
    ->  IO ()

-- | Makes a copy of a complex float array.
foreign import ccall unsafe "convertComplexFloatArrayToComplexDoubleArray" c_convertComplexFloatArrayToComplexDoubleArray ::
        Ptr (Complex Float)
    ->  Ptr (Complex Double)
    ->  Int
    ->  IO ()


-- | Using the supplied sample format, converts a complex float vector into its
--   associated bytestring.
serializeOutput :: CL.SampleFormat -> VST.Vector (Complex Float) -> B.ByteString
serializeOutput CL.SampleComplexDouble signal = runCBasedSerializer (sizeOf(undefined::Complex Double)) c_convertComplexFloatArrayToComplexDoubleArray signal
serializeOutput CL.SampleComplexFloat signal = runCBasedSerializer (sizeOf(undefined::Complex Float)) c_convertFloatArrayToFloatArray signal
serializeOutput CL.SampleComplexSigned16 signal =  runCBasedSerializer (sizeOf(undefined::Word32)) c_convertComplexFloatArrayToComplexSigned16 signal
serializeOutput CL.SampleComplexToDoubleMag signal = runCBasedSerializer (sizeOf(undefined::Double)) c_convertComplexFloatArrayToDoubleMag signal
serializeOutput CL.SampleComplexToFloatMag signal = runCBasedSerializer (sizeOf(undefined::Float)) c_convertComplexFloatArrayToFloatMag signal
serializeOutput CL.SampleComplexToSigned16Mag signal = runCBasedSerializer (sizeOf(undefined::Word16)) c_convertComplexFloatArrayToSigned16Mag signal


{- ### C Functions for deserialization ### -}

-- | Converts a double array into a float array
foreign import ccall unsafe "convertDoubleArrayToFloatArray" c_convertDoubleArrayToFloatArray ::
        CString
    ->  Ptr (Complex Float)
    ->  Int
    ->  IO ()


-- | Converts a float array to a float array. Note, size must be doubled.
foreign import ccall unsafe "convertComplexFloatArrayToComplexFloatArray" c_convertComplexFloatArrayToComplexFloatArray ::
        CString
    ->  Ptr (Complex Float)
    ->  Int
    ->  IO ()


-- | Converts a signed 16 array a float array. Note, the size must be doubled.
foreign import ccall unsafe "convertSigned16ArrayToFloatArray" c_signed16ArrayToComplexFloatArray ::
        CString
    ->  Ptr (Complex Float)
    ->  Int
    ->  IO ()


-- | Runs a C Based serializer that converts a Complex Float Unboxed Vector
--   into a ByteString.
{-# NOINLINE runCBasedSerializer #-}
runCBasedSerializer :: Int -> (Ptr (Complex Float)
    ->  Ptr a
    ->  Int
    ->  IO ()) -> VST.Vector (Complex Float) -> B.ByteString
runCBasedSerializer typeSize c_func signal = unsafeDupablePerformIO $ do
        outputPtr <- mallocBytes arraySize
        withForeignPtr sourcePtr $ \sPtr -> do
            c_func sPtr outputPtr (VST.length signal)
            unsafePackCStringFinalizer (castPtr outputPtr) arraySize (free outputPtr)

    where sourcePtr = fst $ VST.unsafeToForeignPtr0 signal
          arraySize = typeSize * VST.length signal


-- | Deserializes a ByteString into a complex float vector. Note, only supports
--   deserializing complex types. Magnitude types are not supported.
{-# NOINLINE deserializeInput #-}
deserializeInput :: CL.SampleFormat -> B.ByteString -> VST.Vector (Complex Float)
deserializeInput CL.SampleComplexDouble bs = unsafeDupablePerformIO action
    where action = deserializeInputHelper arrayLength bs c_convertDoubleArrayToFloatArray
          arrayLength = quot (B.length bs) (sizeOf(undefined::Complex Double))

deserializeInput CL.SampleComplexFloat bs = unsafeDupablePerformIO action
    where action = deserializeInputHelper arrayLength bs c_convertComplexFloatArrayToComplexFloatArray
          arrayLength = quot (B.length bs) (sizeOf(undefined::Complex Float))

deserializeInput CL.SampleComplexSigned16 bs = unsafeDupablePerformIO action
    where action = deserializeInputHelper arrayLength bs c_signed16ArrayToComplexFloatArray
          arrayLength = quot (B.length bs) (sizeOf(undefined::Word32))

deserializeInput _ _  = VST.empty


-- | Wrapper around common functionlaity used by deserializeInput.
{-# NOINLINE deserializeInputHelper #-}
deserializeInputHelper :: Int -> B.ByteString ->
                           (CString ->  Ptr (Complex Float) -> Int ->  IO ()) ->
                            IO (VST.Vector (Complex Float))
deserializeInputHelper arrayLength bs cFunc = do

        outputPtr <- mallocBytes arrayLengthBytes

        unsafeUseAsCStringLen bs $ \(src, _) ->
                cFunc src outputPtr (2*arrayLength)

        outputForeignPtr <- newForeignPtr finalizerFree outputPtr

        return $ VST.unsafeFromForeignPtr0 outputForeignPtr arrayLength

    where arrayLengthBytes = sizeOf(undefined::Complex Float) * arrayLength


-- | Serializes a complex float.
complexFloatSerializer :: Complex Float -> BP.Put
complexFloatSerializer (real :+ imaginary) = do

    BP.putFloatle real
    BP.putFloatle imaginary

-- | Serializes a Complex Float into a signed 16 with base 1.
--   Faster then complexSigned16Serializer since no division is performed.
complexSigned16SerializerOne :: Complex Float -> BP.Put
complexSigned16SerializerOne (real :+ imaginary) = do
    BP.putInt16host $ convertNumber real
    BP.putInt16host $ convertNumber imaginary

    where convertNumber number = floor $ signum number * abs (minimum [ 32767.0 * number, 32767.0])

-- | Serializes a complex Float as a complex double
complexDoubleSerializer :: Complex Float -> BP.Put
complexDoubleSerializer (real :+ imaginary) = do
    BP.putDoublele  (float2Double real)
    BP.putDoublele  (float2Double imaginary)

File: RadarTools/src/shared/Shared/IO.c
/**
Copyright 2017 Robert Christian Taylor

Licensed under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License.
You may obtain a copy of the License at

    http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software
distributed under the License is distributed on an "AS IS" BASIS,
WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
See the License for the specific language governing permissions and
limitations under the License.
*/

/**
 * @file Routines for converting values to various types. Called from IO.hs.
 * @author Robert Christian Taylor
 * @license Apache-2.0
 * @copyright Robert Christian Taylor
 */

#include <stdint.h>
#include <string.h>

/**
 * Converts an array of signed 16 values to floats.
 * @param source The array of signed 16 values to convert to floats.
 * @param dest The memory location to write the conversion results to.
 * @param size The size of the int16_t array.
 */
void convertSigned16ArrayToFloatArray(int16_t * source, float * dest, int size) 
{
    int i;
    for(i = 0; i < size; i++) 
    {
        dest[i] = (float) source[i] / 32768.0;
    }
}

/**
 * Converts a double array into a float array.
 * @param source Pointer to the array of floats to copy in.
 * @param dest Pointer to the array to write floats into.
 * @param The size of the float array.
 */
void convertDoubleArrayToFloatArray(double * source, float * dest, int size)
{
    int i;
    for(i = 0; i < size; i++) 
    {
        dest[i] = (float)source[i];
    }
}

/**
 * Converts a complex float array into a complex double array.
 * @param source Pointer to the array of floats to read from.
 * @param dest Pointer to the array of doubles to write values to.
 * @param size The size of the array.
 */
void convertComplexFloatArrayToComplexDoubleArray(float * source, double * dest, int size)
{
    int i;
    for(i = 0; i < 2*size; i++)
    {
        dest[i] = (double)source[i];
    }
}

/**
 * Converts a complex float array to a complex signed 16 array.
 * @param source Pointer to complex array to read from.
 * @param dest Pointer to the array to write the signed 16 values to.
 * @param size The size of the array.
 */
void convertComplexFloatArrayToComplexSigned16(float * source, int16_t * dest, int size) 
{
    int i;
    for(i = 0; i < 2*size; i++)
    {
        float result = source[i] * 32767.0;
        if(result > 32767.0) 
        {
            dest[i] = 32767;
        } 
        else if (result < -32767.0) 
        {
            dest[i] = -32767;
        } 
        else
        {
            dest[i] = (int16_t)result;
        }
    }
}

/**
 * Duplicates a complex float array. 
 * @param source The address to copy from.
 * @param dest The address to copy to.
 * @param size The size of the array in complex float.
 */
void convertComplexFloatArrayToComplexFloatArray(float * source, float * dest, int size)
{
    memcpy(dest, source, sizeof(float)*size*2);
}

/**
 * Duplicates a float array
 * @param source The address to copy from.
 * @param dest The address to copy to.
 * @param size The size of the array in floats.
 */
void convertFloatArrayToFloatArray(float * source, float * dest, int size)
{
    memcpy(dest, source, sizeof(float)*size);
}

/**
 * Converts a complex float array to its double magnitude squared (|z|^2).
 * @param source The source array.
 * @param dest The destination array to write the magnitudes to.
 * @param size The size of the array.
 */
void convertComplexFloatArrayToDoubleMag(float * source, double * dest, int size) 
{
    int i;
    for(i = 0; i < size; i++)
    {
        dest[i] = source[2*i]*source[2*i] + source[2*i + 1]*source[2*i + 1];
    }
}

/**
 * Converts a complex float array to its magnitude squared (|z|^2) stored 
 * as floating values.
 * @param source The source array.
 * @param dest The destination array to write the magnitudes into.
 * @param size The size of the array.
 */
void convertComplexFloatArrayToFloatMag(float * source, float * dest, int size)
{
    int i;
    for(i = 0; i < size; i++)
    {
        dest[i] = (float) (source[2*i]*source[2*i] + source[2*i + 1]*source[2*i + 1]);
    }
}

/**
 * Converts a complex float array to an array of the complex number's magnitudes 
 * stored as signed 16 numbers.
 * @param source The location to read from.
 * @param dest The location to write to.
 * @param size The size of the array.
 */
void convertComplexFloatArrayToSigned16Mag(float * source, int16_t * dest, int size)
{
    int i;
    for(i = 0; i < size; i++)
    {
        float result = (source[2*i]*source[2*i] + source[2*i + 1]*source[2*i + 1]) * 32767.0;
        
        if(result > 32767.0) 
        {
            dest[i] = 32767;
        } 
        else if (result < -32767.0) 
        {
            dest[i] = -32767;
        } 
        else
        {
            dest[i] = (int16_t)result;
        }
    }
}
